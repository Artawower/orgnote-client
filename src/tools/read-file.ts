interface FileInfo {
  filePath: string[];
  content: string;
}

export const readFileText = (file: File): Promise<string> => {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onload = () => {
      resolve(reader.result as string);
    };
    reader.onerror = () => {
      reject(reader.error);
    };
    reader.readAsText(file);
  });
};

export const readFile = async (fileEntry: FileSystemEntry): Promise<File> => {
  return new Promise((resolve, reject) => {
    (fileEntry as FileSystemFileEntry).file(resolve, reject);
  });
};

export const readOrgFile = async (
  entry: FileSystemEntry | File
): Promise<FileInfo> => {
  const file = entry instanceof File ? entry : await readFile(entry);
  const fileText = await readFileText(file);
  const path =
    entry instanceof File
      ? file.webkitRelativePath || file.name
      : entry.fullPath;

  const filePath = path.split('/').filter((p) => !!p);

  return {
    filePath,
    content: fileText,
  };
};
