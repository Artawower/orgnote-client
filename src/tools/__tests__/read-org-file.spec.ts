import { readOrgFile } from '../read-file';

describe('read-org-file', () => {
  it('Should read org file by provided entry without start slash', () => {
    const f: File = new File(['#+TITLE: some org file'], 'test.org', {
      type: 'text/org',
    });

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (f as any).webkitRelativePath = 'dir/test.org';

    expect(readOrgFile(f)).resolves.toMatchSnapshot();
  });

  it('Should read org file by provided entry with start slash', () => {
    const f: File = new File(['#+TITLE: some org file'], 'test.org', {
      type: 'text/org',
    });

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (f as any).webkitRelativePath = '/dir/nested/test.org';

    expect(readOrgFile(f)).resolves.toMatchSnapshot();
  });

  it('Should read org file by provided entry with single path', () => {
    const f: File = new File(['#+TITLE: some org file'], 'test.org', {
      type: 'text/org',
    });

    expect(readOrgFile(f)).resolves.toMatchSnapshot();
  });
});
