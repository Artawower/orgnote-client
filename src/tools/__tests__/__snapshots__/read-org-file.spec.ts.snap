// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`read-org-file Should read org file by provided entry without start slash 1`] = `
{
  "content": "#+TITLE: some org file",
  "filePath": [
    "dir",
    "test.org",
  ],
}
`;

exports[`read-org-file Should read org file by provided entry with start slash 1`] = `
{
  "content": "#+TITLE: some org file",
  "filePath": [
    "dir",
    "nested",
    "test.org",
  ],
}
`;

exports[`read-org-file Should read org file by provided entry with single path 1`] = `
{
  "content": "#+TITLE: some org file",
  "filePath": [
    "test.org",
  ],
}
`;
